assets/js/components/KeyMetrics/key-metrics-widgets.js

```diff
@@ -45,6 +45,38 @@ import {
 } from '../../googlesitekit/datastore/user/constants';
 import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
 import { isFeatureEnabled } from '../../features';
+import { MODULES_ANALYTICS_4 } from '../../modules/analytics-4/datastore/constants';
+
+/**
+ * Determines whether to display a widget that reruires custom dimensionsin the key
+ * metrics selection panel.
+ *
+ * All widgets are displayed in authenticated dashboard. However, in view only dashboard,
+ * widgets that require custom dimensions will only be displayed if the required custom
+ * dimensions are available in the shared property.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Function} select              Data store select function.
+ * @param {boolean}  isViewOnlyDashboard Whether the current dashboard is view only.
+ * @return {boolean} Whether to display the widget.
+ */
+function shouldDisplayWidgetWithCustomDimensions(
+	select,
+	isViewOnlyDashboard
+) {
+	if ( ! isFeatureEnabled( 'newsKeyMetrics' ) ) {
+		return false;
+	}
+
+	if ( ! isViewOnlyDashboard ) {
+		return true;
+	}
+
+	return select( MODULES_ANALYTICS_4 ).hasCustomDimensions(
+		this.requiredCustomDimensions
```

Comment:
What's with the `this.` usage here? I'm a bit confused about how this works 😅

Reply:
This is used a little down the same file in `KM_ANALYTICS_LEAST_ENGAGING_PAGES` and `KM_ANALYTICS_VISIT_LENGTH` widgets as  the callback for `displayInList`. I extracted the function so there wouldn't be repeated logic, and the widget objects has the `requiredCustomDimensions` property. Do you think it will be better to make the function pure and pass the `requiredCustomDimensions` inline (still with `this`, so the dimensions are not repeated)?

Reply:
As discussed on slack, the function is directly attached to the widget tile object as a method, and the object themselves has the `requiredCustomDimensions` property , hence I was using this , although it looks sketchy, it avoids some duplication. This is a bit harder to spot in the `diff`. 

We could inline the function call, or pass the custom dimensions inline instead of using `this`. but there is a risk of the logic or the custom dimensions getting out of sync, especially when we add more widget tiles that has more custom dimensions. 

I've added a comment on the function to add further clarification.

Cheers!